File: 03 有环链表II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow=head;
        ListNode* fast=head;
        if(head==NULL||head->next==NULL)
            //注意分析临界情况 (1)成环单节点; (2)不成环单节点; (3)成环2节点; (4)不成环2节点
            return NULL;
        while(slow!=NULL&&fast!=NULL&&fast->next!=NULL){ //判断是否成环
            slow=slow->next;
            fast=fast->next->next;
            if(slow==fast)
                break;
        }
        if(slow==fast){
            slow=head;
            while(slow!=fast&&fast!=NULL&&fast->next!=NULL){ //找到环的起点
                slow=slow->next;
                fast=fast->next;
                if(slow==fast)
                    break;
            }
            return slow;
        }else{
            return NULL;
        }
 
    }
};

File: 03有环链表II.py
"""
给定一个链表，返回链表开始入环的第一个节点。 如果链表无环，则返回 null。
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 双指针，快指针每次比慢指针多走1，当两指针第一次相遇时，则有环，快指针比慢指针多走了环的长度M
# 将快指针重新指向头，快慢指针每次走一步，第二次相遇时，则指的是入环的第一个节点
class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        fast = head
        slow = head
        while fast is not None and fast.next is not None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                fast = head
                while fast is not slow:
                    fast = fast.next
                    slow = slow.next
                return fast
        return None
